### DISCUSSION TITLE
View を拡張したい場合は原則として extension を使用し、状態保持が必要な場合のみ `ViewModifier` を実装する。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#31] [Ideas]
## Overview

SwiftUI では View を拡張する基本機能として Modifier というものが用意されている。

それは SwiftUI を使っていれば誰でも使用するであろう、任意の View に `.xxx()` と書くあれのことだ。

```swift
Text("Hello")
    .border(.blue)
    .padding()
```

ここでは `.border()` と `.padding()` が Modifier ということになる。（おそらく）既存の View を変更（Modify）するので Modifier と呼ばれている。

さて、自分でこのような View を拡張する処理を実装したい場合どうしたらよいだろうか。大きく2つの手段が存在する。

1. View の extension にメソッドを実装する。
2. [ViewModifier](https://developer.apple.com/documentation/swiftui/viewmodifier) プロトコルを実装した独自の Modifier を定義する。

指定された `width` で青い枠線を追加する場合、それぞれ以下のように定義できる。

```swift
// 💡 1. Use extension
extension View {
    func blueBorder(width: CGFloat) -> some View {
        self.border(.blue, width: width)
    }
}

// 💡 2. Define original modifier
struct BlueBorderModifier: ViewModifier {
    let width: CGFloat
    
    func body(content: Content) -> some View {
        content.border(.blue, width: width)
    }
}

struct ContentView: View {
    var body: some View {
        VStack(spacing: 8) {
            Text("Hello.")
                .blueBorder(width: 2) // 💡 1. Use extension method
            
            Text("Hello.")
                .modifier(BlueBorderModifier(width: 2)) // 💡 2. Use original modifier
        }
    }
}
```

これらはコードスタイルや記述量を無視すれば、どちらも同じ処理に見えるかも知れない。実際のところそのとおりで、どちらの実装方法を採用しても同じ表示結果が得られる。

<img width="126" alt="image" src="https://user-images.githubusercontent.com/2990285/164567548-bc00cde1-0b5b-4fa4-8f89-00cd4b90d9ae.png">

ではどちらを利用すべきなのだろうか？

おそらく多くの人が直感的に予想しているだろうとおり、以下の2点の理由から 1. の extension を使った方がベターなはずだ。

- 記述量が少なくて済む。
- コード補完が快適にできる。（`.modifier()`を経由する必要がない）

それでは 2. の独自 Modifier を定義したいケースというのはいつだろう？

実は `ViewModifier` は `View` と同じように `@State` などを利用して **状態を保持する View を作成することが可能** になっている。

以下は指定された View の下部に、ON/OFF用の [Toggle](https://developer.apple.com/documentation/swiftui/toggle) を設置し、ON/OFF の状態によって枠線（`border`）の色を変化させる Modifier となる。

```swift
struct SwitchBorderModifier: ViewModifier {
    @State var isOn = false // ✅ Can hold state as like `View`.
    
    func body(content: Content) -> some View {
        VStack {
            content.border(isOn ? .red : .blue) // 💡 Switch red and blue.
            Toggle("Border", isOn: $isOn)
        }
    }
}
```

このように状態を保持したい View 拡張を提供したい場合は extension では実現できないため、そのようなケースでのみ独自の `ViewModifier` を定義する必要がある。

## Discussion

`ViewModifier` で定義した場合も、`View` の extension にショートカット用のメソッドを定義しておくと、extension と同じように快適なコード補完が可能となる。

```swift
extension View {
    // ✅ Shortcut method for `SwitchBorderModifier`
    func switchBorder() -> some View {
        self.modifier(SwitchBorderModifier())
    }
}

struct ContentView: View {
    var body: some View {
        Text("Hello.").switchBorder() // ✅ Easy to use.
    }
}
```

特別な事情がない限り、これは常に ViewModifier と一緒に定義するようにすべきだろう。そうしておけば実装する際に利用方法に迷うこともないだろうし、コードスタイルとしても統一される。

## Reference

- [ViewModifier](https://developer.apple.com/documentation/swiftui/viewmodifier)

**Comment by uhooi** [comment]
ちょうど悩んでいたので、非常に参考になりました…！

ただ extension の使用時、以下のエラーが発生することがありました。

```swift
private extension View {
  func roundedBorder(shouldShow: Bool) -> some View {
    if shouldShow {
      return self.overlay(
        RoundedRectangle(cornerRadius: 8)
          .stroke(.blue, lineWidth: 4)
      )
    } else {
      return self
    }
  }
}
```

```
function declares an opaque return type, but the return statements in its body do not have matching underlying types
```

`@ViewBuilder` を付けて `return` を省略するとビルドが通るようになりました。

```diff
private extension View {
+ @ViewBuilder
  func roundedBorder(shouldShow: Bool) -> some View {
    if shouldShow {
-     return self.overlay(
+     self.overlay(
        RoundedRectangle(cornerRadius: 8)
          .stroke(.blue, lineWidth: 4)
      )
    } else {
-     return self
+     self
    }
  }
}
```

まだちゃんと理解していないのですが、ViewBuilder についても触れていると初心者（私）に嬉しいです…！

**Reply by uhooi** [reply]
ref: https://twitter.com/the_uhooi/status/1544197851852701697?s=20

**Reply by YusukeHosonuma** [reply]
タイトルが微妙なのですが、 `@VIewBuilder` については以下で説明していました 🙏 

本当に必要な場合を除き AnyView を避ける。
https://github.com/YusukeHosonuma/Effective-SwiftUI/discussions/7

個人的に `@ViewBuilder` については `Learn` に該当するものと考えていたので、（`Effective Use` としての）トピックは設けないで良いかと思っていたのですが、今後ちょっと検討してみたいと思いました。

**Reply by uhooi** [reply]
ああ、思いっきり書いてありましたね、、
最初に読んだときは `@ViewBuilder` の意味がわかっていなかったのですが、if 文などがあって返す型がビルド時に確定しないときに必要なのですね。
完全に理解しました！

**Reply by uhooi** [reply]
改めて自分のコメントを読み返したら、 `@ViewBuilder` は `Learn` に該当すると思うので、ここで明記するものではないと思いました :pray:
